build-system/tasks/helpers.js

```diff
@@ -150,6 +150,20 @@ async function bootstrapThirdPartyFrames(options) {
  * @param {!Object} options
  */
 async function compileCoreRuntime(options) {
+  async function watchFunc() {
+    const bundleComplete = await doBuildJs(jsBundles, 'amp.js', options);
+    if (options.onWatchBuild) {
+      options.onWatchBuild(bundleComplete);
+    }
+  }
+
+  if (options.watch) {
+    fileWatch('src/**/*.js').on(
+      'change',
+      debounce(watchFunc, watchDebounceDelay)
+    );
+  }
+
```

Comment:
As written, I suspect this PR will result in a double-build of `v0.js` or `amp.js` when you run `gulp` and then edit `src/amp.js`. This is because `dobuildJs()` calls `compileJs()`, which calls its minified / unminfied equivalents, which in turn sets up a watcher for `src/amp.js`.

Extensions solve this problem in `buildExtension()` by setting up a top-level watcher for the extension directory, and then setting `options.watch` to false before calling the inner build functions so they don't create a second watcher.

https://github.com/ampproject/amphtml/blob/319372d708ffff4ec7ff990603edd0b9f9101286/build-system/tasks/extension-helpers.js#L435-L446

Might have to do something like that here.

Reply:
Actually, its even weirder. As written, this PR _only_ works when `amp.js` is modified, which happens to be the only file I had tested 😬 .  That is because `gulp` default task starts the lazy server which does not call into `compileCoreRuntime`. 

Fix incoming